### PR TITLE
API Server: migrations, pagination, authentication, search

### LINKED ISSUE #30: API backend: Authentication
### What is the expected behavior?
Add an authentication system for requesting the remote API.

### Questions
- How do we want to manage the authentication with the remote API ?
Using the IBMQ experience like Qiskit does or something else ... ?
- Do we want to do that for the QAMP or for after ?

### LINKED ISSUE #34: Backend: search
### What is the expected enhancement?

Implement search functionality. Let's start with search by title and description.

### Details

In client:
```python
backend.list(query="<my-seach-term>")
```

API backend: `/trials/?query=<bla-bla-my-search-term>`
```python
Trial.objects.filter(title__contains="<my_search_term>" or description__contain="<my_search_term>")
```

LocalBackend: TBD

### PR DESCRIPTION
**Pull request by IceKhan13**
### Summary

Improvements to API server
- authentication
- migrations
- pagination
- search

### Details and comments

**Authentication**:

Get token:

```shell
curl --location --request POST 'http://127.0.0.1:8000/api/token/' \
--header 'Content-Type: application/json' \
--data-raw '{
    "username": "admin",
    "password": 123
}'
```

Response:
```json
{
    "refresh": "...",
    "access": "..."
}
```

Use token:

```shell
curl --location --request GET 'http://127.0.0.1:8000/api/trials/' \
--header 'Authorization: Bearer <ACCESS_TOKEN>'
```

**Search and pagination**

```shell
curl --location --request GET 'http://127.0.0.1:8000/api/trials/?query=term&offset=0&limit=20' \
--header 'Authorization: Bearer <ACCESS_TOKEN>'
```

Closes #34 
Closes #30

### CODE REVIEW

File: docker-compose.yml
Comment (mickahell) [inline docker-compose.yml:1]: We can variabilize it with the same vars as [.envrc](https://github.com/IceKhan13/purplecaffeine/blob/main/.envrc) in order to have only one place to put variable in.

```bash
POSTGRES_DB: ${DB_NAME:-purplecaffeine}
```

**Review by mickahell — CHANGES_REQUESTED** [review]
Does not  the migrations files generated automatically by running :
`python3 manage.py makemigrations core` ?

Else just a stuff for compose and that's good to me :)

File: api_server/core/views.py [hunk 1]
```diff
@@ -11,5 +14,15 @@ class TrialViewSet(viewsets.ModelViewSet):
     ViewSet for Trial model
     """
 
-    queryset = Trial.objects.all()  # pylint: disable=no-member
+    permission_classes = [permissions.IsAuthenticated]
```
Comment (mickahell) [inline api_server/core/views.py:17]: SO what kind of token is it ? Can you add a curl example in the readme :)

Comment (IceKhan13) [reply api_server/core/views.py:17]: I updated description with examples :)

File: docker-compose.yml
Comment (IceKhan13) [reply docker-compose.yml:1]: Done :)

**Comment by IceKhan13** [comment]
@mickahell we need to keep track of migrations each time we are doing changes in application models. Therefore migrations should be generated and always tracked

**Review by mickahell — APPROVED** [review]
Good to me 🚀 I create the issue #45 to update the client with your changes :)

**Pull request merged**